File: .skills/openclaw-skills/skills/0xnyk/xint/lib/collections.ts
/**
 * lib/collections.ts — xAI Collections Knowledge Base integration
 *
 * Uses xAI's APIs:
 * - api.x.ai (XAI_API_KEY): files, search
 * - management-api.x.ai (XAI_MANAGEMENT_API_KEY): collections management
 */

import { readFileSync } from "fs";
import { join } from "path";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface Collection {
  id: string;
  name: string;
  description?: string;
  created_at?: string;
}

export interface FileUploadResponse {
  id: string;
  filename: string;
  purpose: string;
  status: string;
}

export interface DocumentSearchResult {
  id: string;
  content?: string;
  score?: number;
  metadata?: Record<string, unknown>;
}

export interface CollectionsListResponse {
  data: Collection[];
}

export interface DocumentSearchResponse {
  results: DocumentSearchResult[];
}

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const API_BASE = "https://api.x.ai/v1";
const MGMT_BASE = "https://management-api.x.ai/v1";

/** Normalize xAI API response to our Collection interface (handles collection_id/collection_name fields) */
function normalizeCollection(raw: unknown): Collection {
  const r = raw as Record<string, unknown>;
  return {
    id: (r.id || r.collection_id || "") as string,
    name: (r.name || r.collection_name || "") as string,
    description: (r.description || undefined) as string | undefined,
    created_at: (r.created_at || undefined) as string | undefined,
  };
}

// ---------------------------------------------------------------------------
// Auth
// ---------------------------------------------------------------------------

function readEnvKey(varName: string): string | undefined {
  if (process.env[varName]) return process.env[varName];
  // Check .env.local first (overrides), then .env
  for (const file of [".env.local", ".env"]) {
    try {
      const envFile = readFileSync(join(import.meta.dir, "..", file), "utf-8");
      const re = new RegExp(`${varName}=["']?([^"'\\n]+)`);
      const match = envFile.match(re);
      if (match && !match[1].startsWith("your-")) return match[1];
    } catch {}
  }
  return undefined;
}

function getXaiKey(): string {
  const key = readEnvKey("XAI_API_KEY");
  if (key) return key;
  throw new Error("XAI_API_KEY not found. Set it in your environment or in .env");
}

function getMgmtKey(): string {
  const key = readEnvKey("XAI_MANAGEMENT_API_KEY");
  if (key) return key;

  throw new Error("XAI_MANAGEMENT_API_KEY not found. Set it in your environment or in .env");
}

// ---------------------------------------------------------------------------
// Collections API (Management)
// ---------------------------------------------------------------------------

/**
 * List all collections.
 */
export async function collectionsList(): Promise<CollectionsListResponse> {
  const key = getMgmtKey();

  const res = await fetch(`${MGMT_BASE}/collections`, {
    headers: {
      Authorization: `Bearer ${key}`,
    },
  });

  if (!res.ok) {
    const text = await res.text();
    throw new Error(`xAI API error (${res.status}): ${text.slice(0, 500)}`);
  }

  const data = await res.json() as Record<string, unknown>;
  const rawList = (data.data || data.collections || []) as unknown[];
  return { data: rawList.map(normalizeCollection) };
}

/**
 * Create a collection.
 */
export async function collectionsCreate(
  name: string,
  description: string = ""
): Promise<Collection> {
  const key = getMgmtKey();

  const body: Record<string, string> = { collection_name: name };
  if (description) {
    body.description = description;
  }

  const res = await fetch(`${MGMT_BASE}/collections`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${key}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(body),
  });

  if (!res.ok) {
    const text = await res.text();
    throw new Error(`xAI API error (${res.status}): ${text.slice(0, 500)}`);
  }

  const data = await res.json() as Record<string, unknown>;
  return normalizeCollection(data.collection || data.data || data);
}

/**
 * Ensure a collection exists (create if not present).
 */
export async function collectionsEnsure(
  name: string,
  description: string = ""
): Promise<{ collection: Collection; created: boolean }> {
  const list = await collectionsList();

  // Check if exists
  const existing = (list.data || []).find((c) => c.name === name);
  if (existing) {
    return { collection: existing, created: false };
  }

  // Create new
  const created = await collectionsCreate(name, description);
  return { collection: created, created: true };
}

/**
 * Add an already-uploaded document to a collection by ID.
 */
export async function collectionsAddDocument(
  collectionId: string,
  documentId: string
): Promise<unknown> {
  const key = getMgmtKey();

  // Management API requires multipart form for this endpoint
  const proc = Bun.spawn([
    "curl", "-s",
    "-X", "POST",
    `${MGMT_BASE}/collections/${collectionId}/documents`,
    "-H", `Authorization: Bearer ${key}`,
    "-F", `document_id=${documentId}`,
  ], { stdout: "pipe", stderr: "pipe" });

  const output = await new Response(proc.stdout).text();
  await proc.exited;

  if (!output.trim()) {
    throw new Error("Empty response from xAI management API");
  }

  const data = JSON.parse(output.trim());
  if (data.code || data.message) {
    throw new Error(`xAI API error: ${data.message || output.slice(0, 300)}`);
  }

  return data;
}

/**
 * Upload a file directly to a collection (file + metadata in one call).
 */
export async function collectionsUploadDocument(
  collectionId: string,
  filePath: string,
  name: string,
  contentType: string = "text/markdown"
): Promise<unknown> {
  const key = getMgmtKey();

  const proc = Bun.spawn([
    "curl", "-s",
    "-X", "POST",
    `${MGMT_BASE}/collections/${collectionId}/documents`,
    "-H", `Authorization: Bearer ${key}`,
    "-F", `file=@${filePath}`,
    "-F", `data=@${filePath}`,
    "-F", `name=${name}`,
    "-F", `content_type=${contentType}`,
  ], { stdout: "pipe", stderr: "pipe" });

  const output = await new Response(proc.stdout).text();
  await proc.exited;

  if (!output.trim()) {
    throw new Error("Empty response from xAI management API");
  }

  const data = JSON.parse(output.trim());
  if (data.code || data.message) {
    throw new Error(`xAI API error: ${data.message || output.slice(0, 300)}`);
  }

  return data;
}

// ---------------------------------------------------------------------------
// Files API
// ---------------------------------------------------------------------------

/**
 * Upload a file to xAI.
 */
export async function filesUpload(
  filePath: string,
  purpose: string = "assistants"
): Promise<FileUploadResponse> {
  const key = getXaiKey();

  const filename = filePath.split("/").pop() || "file";

  const proc = Bun.spawn([
    "curl", "-s",
    "-X", "POST",
    `${API_BASE}/files`,
    "-H", `Authorization: Bearer ${key}`,
    "-F", `file=@${filePath};filename=${filename}`,
    "-F", `purpose=${purpose}`,
  ], { stdout: "pipe", stderr: "pipe" });

  const output = await new Response(proc.stdout).text();
  await proc.exited;

  if (!output.trim()) {
    throw new Error("Empty response from xAI files API");
  }

  const data = JSON.parse(output.trim()) as Record<string, unknown>;
  if ((data as any).error || (data as any).code) {
    throw new Error(`xAI API error: ${output.slice(0, 500)}`);
  }

  const result = (data.data || data) as FileUploadResponse;
  return result;
}

// ---------------------------------------------------------------------------
// Documents Search API
// ---------------------------------------------------------------------------

/**
 * Search documents in collections.
 */
export async function documentsSearch(
  query: string,
  collectionIds: string[] = [],
  topK: number = 8
): Promise<DocumentSearchResponse> {
  const key = getXaiKey();

  const body: Record<string, unknown> = {
    query,
    top_k: topK,
    source: "user_upload",
  };

  if (collectionIds.length > 0) {
    body.collection_ids = collectionIds;
  }

  const res = await fetch(`${API_BASE}/documents/search`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${key}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(body),
  });

  if (!res.ok) {
    const text = await res.text();
    throw new Error(`xAI API error (${res.status}): ${text.slice(0, 500)}`);
  }

  return res.json() as Promise<DocumentSearchResponse>;
}

// ---------------------------------------------------------------------------
// CLI Handler
// ---------------------------------------------------------------------------

export async function cmdCollections(args: string[]): Promise<void> {
  const subcommand = args[0] || "help";

  try {
    switch (subcommand) {
      case "list":
        await cmdList();
        break;
      case "create":
        await cmdCreate(args.slice(1));
        break;
      case "ensure":
        await cmdEnsure(args.slice(1));
        break;
      case "add-document":
      case "add-doc":
        await cmdAddDocument(args.slice(1));
        break;
      case "upload":
        await cmdUpload(args.slice(1));
        break;
      case "search":
        await cmdSearch(args.slice(1));
        break;
      case "sync-dir":
      case "sync":
        await cmdSyncDir(args.slice(1));
        break;
      case "help":
      case "--help":
      case "-h":
        printHelp();
        break;
      default:
        console.error(`Unknown subcommand: ${subcommand}`);
        printHelp();
        process.exit(1);
    }
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    console.error(`Error: ${msg}`);
    process.exit(1);
  }
}

async function cmdList(): Promise<void> {
  const res = await collectionsList();
  console.log(JSON.stringify(res, null, 2));
}

async function cmdCreate(args: string[]): Promise<void> {
  const name = getFlag(args, "--name");
  const description = getFlag(args, "--description") || "";

  if (!name) {
    console.error("Error: --name required");
    process.exit(1);
  }

  const collection = await collectionsCreate(name, description);
  console.log(JSON.stringify(collection, null, 2));
}

async function cmdEnsure(args: string[]): Promise<void> {
  const name = getFlag(args, "--name");
  const description = getFlag(args, "--description") || "";

  if (!name) {
    console.error("Error: --name required");
    process.exit(1);
  }

  const result = await collectionsEnsure(name, description);
  console.log(JSON.stringify(result, null, 2));
}

async function cmdAddDocument(args: string[]): Promise<void> {
  const collectionId = getFlag(args, "--collection-id");
  const documentId = getFlag(args, "--document-id");

  if (!collectionId || !documentId) {
    console.error("Error: --collection-id and --document-id required");
    process.exit(1);
  }

  const result = await collectionsAddDocument(collectionId, documentId);
  console.log(JSON.stringify(result, null, 2));
}

async function cmdUpload(args: string[]): Promise<void> {
  const path = getFlag(args, "--path");
  const purpose = getFlag(args, "--purpose") || "kb_sync";

  if (!path) {
    console.error("Error: --path required");
    process.exit(1);
  }

  const result = await filesUpload(path, purpose);
  console.log(JSON.stringify(result, null, 2));
}

async function cmdSearch(args: string[]): Promise<void> {
  const query = getFlag(args, "--query");
  const collectionIdsStr = getFlag(args, "--collection-ids") || "";
  const topK = parseInt(getFlag(args, "--top-k") || "8");

  if (!query) {
    console.error("Error: --query required");
    process.exit(1);
  }

  const collectionIds = collectionIdsStr
    .split(",")
    .map((s) => s.trim())
    .filter(Boolean);

  const result = await documentsSearch(query, collectionIds, topK);
  console.log(JSON.stringify(result, null, 2));
}

async function cmdSyncDir(args: string[]): Promise<void> {
  const collectionName = getFlag(args, "--collection-name") || getFlag(args, "--name");
  const dir = getFlag(args, "--dir");
  const globs = args.filter((a) => a.startsWith("--glob")).map((a) => a.replace("--glob=", "").replace("--glob", ""));
  const limit = parseInt(getFlag(args, "--limit") || "50");
  const purpose = getFlag(args, "--purpose") || "kb_sync";
  const reportMd = getFlag(args, "--report-md");

  if (!collectionName || !dir) {
    console.error("Error: --collection-name and --dir required");
    process.exit(1);
  }

  const directory = Bun.spawnSync(["ls", dir]);
  if (!directory.success) {
    console.error(`Error: Directory not found: ${dir}`);
    process.exit(1);
  }

  // Ensure collection
  console.log(`Ensuring collection '${collectionName}'...`);
  const { collection, created } = await collectionsEnsure(collectionName, "xint KB sync");
  const collectionId = collection.id;

  if (created) {
    console.log(`Created new collection: ${collectionName}`);
  } else {
    console.log(`Using existing collection: ${collectionName}`);
  }

  // Find files
  const files: string[] = [];
  const globPatterns = globs.length > 0 ? globs : ["*.md"];

  for (const pattern of globPatterns) {
    const result = Bun.spawnSync(["find", dir, "-name", pattern, "-type", "f"]);
    if (result.success && result.stdout) {
      const fileList = new TextDecoder().decode(result.stdout).trim().split("\n");
      files.push(...fileList.filter(Boolean));
    }
  }

  // Dedupe
  const uniqueFiles = [...new Set(files)].slice(0, limit);
  console.log(`Found ${uniqueFiles.length} files matching ${globPatterns.join(", ")}`);

  // Upload and attach
  let uploaded = 0;
  let attached = 0;
  const failures: string[] = [];

  for (const filePath of uniqueFiles) {
    try {
      console.log(`  Uploading ${filePath.split("/").pop()}...`);
      const upRes = await filesUpload(filePath, purpose);
      uploaded++;

      if (collectionId) {
        try {
          await collectionsAddDocument(collectionId, upRes.id);
          attached++;
          console.log(`    Uploaded + attached`);
        } catch (e: any) {
          console.log(`    Uploaded (attach failed: ${e.message})`);
          failures.push(`${filePath}: attach error`);
        }
      }
    } catch (e: any) {
      console.log(`    FAILED: ${e.message}`);
      failures.push(`${filePath}: ${e.message}`);
    }
  }

  // Write report
  const ts = new Date().toISOString();
  const reportPath = reportMd || `${dir}/xai-collections-sync-latest.md`;

  const lines: string[] = [];
  lines.push("# xAI Collections KB Sync");
  lines.push("");
  lines.push(`- Timestamp (UTC): ${ts}`);
  lines.push(`- Collection name: \`${collectionName}\``);
  lines.push(`- Directory: \`${dir}\``);
  lines.push(`- Globs: \`${globPatterns.join(", ")}\``);
  lines.push(`- Limit: ${limit}`);
  lines.push("");
  lines.push("## Summary");
  lines.push("");
  lines.push(`- Files considered: ${uniqueFiles.length}`);
  lines.push(`- Upload attempts: ${uploaded}`);
  lines.push(`- Attach attempts: ${attached}`);
  lines.push(`- Failures: ${failures.length}`);
  lines.push("");

  if (failures.length > 0) {
    lines.push("## Failures");
    lines.push("");
    for (const f of failures.slice(0, 50)) {
      lines.push(`- ${f}`);
    }
    lines.push("");
  }

  await Bun.write(reportPath, lines.join("\n"));

  console.log(`\nSync complete: ${uploaded} uploaded, ${attached} attached, ${failures.length} failures`);
  console.log(`Report: ${reportPath}`);
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function getFlag(args: string[], flag: string): string | undefined {
  const idx = args.indexOf(flag);
  if (idx >= 0 && idx + 1 < args.length) {
    return args[idx + 1];
  }
  // Also check flag=value format
  for (const arg of args) {
    if (arg.startsWith(flag + "=")) {
      return arg.slice(flag.length + 1);
    }
  }
  return undefined;
}

function printHelp(): void {
  console.log(`
xAI Collections Knowledge Base

Usage: xint collections <subcommand> [options]

Subcommands:
  list                                List all collections
  create --name <name> [--description <desc>]  Create a collection
  ensure --name <name> [--description <desc>]  Create if doesn't exist
  add-document --collection-id <id> --document-id <id>  Attach document
  upload --path <file> [--purpose <p>]        Upload a file
  search --query <q> [--collection-ids <ids>]  Search documents
  sync-dir --collection-name <n> --dir <path>  Sync directory to collection

Options:
  --collection-name <name>       Collection name
  --description <desc>           Collection description
  --collection-id <id>           Collection ID
  --document-id <id>             Document/file ID
  --path <file>                  File path to upload
  --purpose <purpose>            Upload purpose (default: kb_sync)
  --query <query>                Search query
  --collection-ids <ids>          Comma-separated collection IDs
  --top-k <N>                     Number of results (default: 8)
  --dir <path>                   Directory to sync
  --glob <pattern>                File glob pattern (default: *.md)
  --limit <N>                    Max files to upload (default: 50)
  --report-md <path>             Report output path

Env vars:
  XAI_API_KEY                     File upload + document search
  XAI_MANAGEMENT_API_KEY          Collections management
`);
}

File: .skills/openclaw-skills/skills/0xnyk/xint-rs/src/commands/collections.rs
use anyhow::{bail, Result};
use std::path::PathBuf;

use crate::api::xai;
use crate::cli::CollectionsArgs;
use crate::config::Config;
use crate::format;

pub async fn run(args: &CollectionsArgs, config: &Config) -> Result<()> {
    let http = reqwest::Client::new();
    let parts: Vec<String> = args.subcommand.clone().unwrap_or_default();
    let sub = parts.first().map(|s| s.as_str()).unwrap_or("help");

    match sub {
        "list" => cmd_list(&http, config).await,
        "create" => cmd_create(&http, config, &parts).await,
        "ensure" => cmd_ensure(&http, config, &parts).await,
        "add-document" | "add-doc" => cmd_add_document(&http, config, &parts).await,
        "upload" => cmd_upload(&http, config, &parts).await,
        "search" => cmd_search(&http, config, &parts, args).await,
        "sync-dir" | "sync" => cmd_sync_dir(&http, config, args).await,
        "help" | "--help" | "-h" => {
            print_help();
            Ok(())
        }
        _ => {
            eprintln!("Unknown collections subcommand: {sub}");
            print_help();
            Ok(())
        }
    }
}

async fn cmd_list(http: &reqwest::Client, config: &Config) -> Result<()> {
    let key = config.require_xai_management_key()?;
    let res = xai::collections_list(http, key).await?;
    format::print_json_pretty_filtered(&res)?;
    Ok(())
}

async fn cmd_create(http: &reqwest::Client, config: &Config, parts: &[String]) -> Result<()> {
    let key = config.require_xai_management_key()?;
    let name = find_flag(parts, "--name")
        .ok_or_else(|| anyhow::anyhow!("--name required for 'collections create'"))?;
    let desc = find_flag(parts, "--description").unwrap_or_default();
    let res = xai::collections_create(http, key, &name, &desc).await?;
    format::print_json_pretty_filtered(&res)?;
    Ok(())
}

async fn cmd_ensure(http: &reqwest::Client, config: &Config, parts: &[String]) -> Result<()> {
    let key = config.require_xai_management_key()?;
    let name = find_flag(parts, "--name")
        .ok_or_else(|| anyhow::anyhow!("--name required for 'collections ensure'"))?;
    let desc = find_flag(parts, "--description").unwrap_or_default();
    let res = xai::collections_ensure(http, key, &name, &desc).await?;
    format::print_json_pretty_filtered(&res)?;
    Ok(())
}

async fn cmd_add_document(http: &reqwest::Client, config: &Config, parts: &[String]) -> Result<()> {
    let key = config.require_xai_management_key()?;
    let collection_id = find_flag(parts, "--collection-id")
        .ok_or_else(|| anyhow::anyhow!("--collection-id required"))?;
    let document_id = find_flag(parts, "--document-id")
        .ok_or_else(|| anyhow::anyhow!("--document-id required"))?;
    let res = xai::collections_add_document(http, key, &collection_id, &document_id).await?;
    format::print_json_pretty_filtered(&res)?;
    Ok(())
}

async fn cmd_upload(http: &reqwest::Client, config: &Config, parts: &[String]) -> Result<()> {
    let api_key = config.require_xai_key()?;
    let path_str =
        find_flag(parts, "--path").ok_or_else(|| anyhow::anyhow!("--path required for upload"))?;
    let purpose = find_flag(parts, "--purpose").unwrap_or_else(|| "kb_sync".to_string());

    let path = PathBuf::from(&path_str);
    if !path.exists() {
        bail!("File not found: {path_str}");
    }

    let res = xai::files_upload(http, api_key, &path, &purpose).await?;
    format::print_json_pretty_filtered(&res)?;
    Ok(())
}

async fn cmd_search(
    http: &reqwest::Client,
    config: &Config,
    parts: &[String],
    args: &CollectionsArgs,
) -> Result<()> {
    let api_key = config.require_xai_key()?;
    let query = find_flag(parts, "--query")
        .ok_or_else(|| anyhow::anyhow!("--query required for search"))?;
    let cids_str = find_flag(parts, "--collection-ids").unwrap_or_default();
    let collection_ids: Vec<String> = cids_str
        .split(',')
        .map(|s| s.trim().to_string())
        .filter(|s| !s.is_empty())
        .collect();
    let top_k = args.top_k;

    let res = xai::documents_search(http, api_key, &collection_ids, &query, top_k).await?;
    format::print_json_pretty_filtered(&res)?;
    Ok(())
}

async fn cmd_sync_dir(
    http: &reqwest::Client,
    config: &Config,
    args: &CollectionsArgs,
) -> Result<()> {
    let api_key = config.require_xai_key()?;
    let mgmt_key = config.require_xai_management_key()?;
    let parts: Vec<String> = args.subcommand.clone().unwrap_or_default();

    let collection_name = find_flag(&parts, "--collection-name")
        .or_else(|| find_flag(&parts, "--name"))
        .ok_or_else(|| anyhow::anyhow!("--collection-name required for sync-dir"))?;
    let dir_str =
        find_flag(&parts, "--dir").ok_or_else(|| anyhow::anyhow!("--dir required for sync-dir"))?;
    let directory = PathBuf::from(&dir_str);
    if !directory.exists() {
        bail!("Directory not found: {dir_str}");
    }

    let globs: Vec<String> = {
        let g = find_all_flags(&parts, "--glob");
        if g.is_empty() {
            vec!["*.md".to_string()]
        } else {
            g
        }
    };
    let limit: usize = find_flag(&parts, "--limit")
        .and_then(|v| v.parse().ok())
        .unwrap_or(50);
    let purpose = find_flag(&parts, "--purpose").unwrap_or_else(|| "kb_sync".to_string());

    let ts = chrono::Utc::now().format("%Y-%m-%dT%H:%M:%SZ").to_string();

    // 1) Ensure collection
    eprintln!("Ensuring collection '{collection_name}'...");
    let ensure_res =
        xai::collections_ensure(http, mgmt_key, &collection_name, "xint KB sync").await?;

    let collection_id = extract_collection_id(&ensure_res);
    if collection_id.is_empty() {
        eprintln!(
            "WARN: could not determine collection_id from response; add-document step may fail."
        );
    }

    // 2) Enumerate files
    let mut paths: Vec<PathBuf> = Vec::new();
    for g in &globs {
        if let Ok(entries) = glob::glob(&directory.join(g).to_string_lossy()) {
            for entry in entries.flatten() {
                if entry.is_file() {
                    paths.push(entry);
                }
            }
        } else {
            // Fallback: manual glob matching
            if let Ok(entries) = std::fs::read_dir(&directory) {
                for entry in entries.flatten() {
                    let path = entry.path();
                    if path.is_file() {
                        let name = path.file_name().and_then(|n| n.to_str()).unwrap_or("");
                        if matches_simple_glob(name, g) {
                            paths.push(path);
                        }
                    }
                }
            }
        }
    }

    // Dedupe
    let mut seen = std::collections::HashSet::new();
    paths.retain(|p| {
        let key = p.to_string_lossy().to_string();
        seen.insert(key)
    });
    paths.sort();
    paths.truncate(limit);

    eprintln!("Found {} files matching globs", paths.len());

    // 3) Upload + attach
    let mut uploaded = 0;
    let mut attached = 0;
    let mut failures: Vec<String> = Vec::new();

    for p in &paths {
        let fname = p.file_name().and_then(|n| n.to_str()).unwrap_or("?");
        eprint!("  Uploading {fname}... ");

        match xai::files_upload(http, api_key, p, &purpose).await {
            Ok(up_res) => {
                uploaded += 1;
                let doc_id = extract_document_id(&up_res);

                if !collection_id.is_empty() && !doc_id.is_empty() {
                    match xai::collections_add_document(http, mgmt_key, &collection_id, &doc_id)
                        .await
                    {
                        Ok(_) => {
                            attached += 1;
                            eprintln!("uploaded + attached");
                        }
                        Err(e) => {
                            eprintln!("uploaded (attach failed: {e})");
                            failures.push(format!("{fname}: attach error: {e}"));
                        }
                    }
                } else {
                    eprintln!("uploaded (no collection_id or doc_id to attach)");
                }
            }
            Err(e) => {
                eprintln!("FAILED: {e}");
                failures.push(format!("{fname}: {e}"));
            }
        }
    }

    // 4) Write report
    let report_path = find_flag(&parts, "--report-md")
        .map(PathBuf::from)
        .unwrap_or_else(|| directory.join("xai-collections-sync-latest.md"));

    let mut lines = Vec::new();
    lines.push("# xAI Collections KB Sync".to_string());
    lines.push(String::new());
    lines.push(format!("- Timestamp (UTC): {ts}"));
    lines.push(format!("- Collection name: `{collection_name}`"));
    lines.push(format!("- Directory: `{}`", directory.display()));
    lines.push(format!("- Globs: `{}`", globs.join(", ")));
    lines.push(format!("- Limit: {limit}"));
    lines.push(String::new());
    lines.push("## Summary".to_string());
    lines.push(String::new());
    lines.push(format!("- Files considered: {}", paths.len()));
    lines.push(format!("- Upload attempts: {uploaded}"));
    lines.push(format!("- Attach attempts: {attached}"));
    lines.push(format!("- Failures: {}", failures.len()));
    lines.push(String::new());
    if !failures.is_empty() {
        lines.push("## Failures".to_string());
        lines.push(String::new());
        for f in &failures[..failures.len().min(50)] {
            lines.push(format!("- {f}"));
        }
        lines.push(String::new());
    }

    if let Some(parent) = report_path.parent() {
        std::fs::create_dir_all(parent)?;
    }
    std::fs::write(&report_path, lines.join("\n").trim_end().to_owned() + "\n")?;

    eprintln!(
        "\nSync complete: {} uploaded, {} attached, {} failures",
        uploaded,
        attached,
        failures.len()
    );
    eprintln!("Report: {}", report_path.display());

    Ok(())
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

fn find_flag(parts: &[String], flag: &str) -> Option<String> {
    for (i, p) in parts.iter().enumerate() {
        if p == flag {
            if let Some(val) = parts.get(i + 1) {
                return Some(val.clone());
            }
        }
        if let Some(rest) = p.strip_prefix(&format!("{flag}=")) {
            return Some(rest.to_string());
        }
    }
    None
}

fn find_all_flags(parts: &[String], flag: &str) -> Vec<String> {
    let mut vals = Vec::new();
    for (i, p) in parts.iter().enumerate() {
        if p == flag {
            if let Some(val) = parts.get(i + 1) {
                vals.push(val.clone());
            }
        }
        if let Some(rest) = p.strip_prefix(&format!("{flag}=")) {
            vals.push(rest.to_string());
        }
    }
    vals
}

fn extract_collection_id(res: &serde_json::Value) -> String {
    // Try collection.id or created.data.id
    if let Some(coll) = res.get("collection") {
        for key in &["id", "collection_id"] {
            if let Some(id) = coll.get(key).and_then(|v| v.as_str()) {
                let s = id.trim();
                if !s.is_empty() {
                    return s.to_string();
                }
            }
        }
    }
    if let Some(created) = res.get("created") {
        if let Some(data) = created.get("data") {
            for key in &["id", "collection_id"] {
                if let Some(id) = data.get(key).and_then(|v| v.as_str()) {
                    let s = id.trim();
                    if !s.is_empty() {
                        return s.to_string();
                    }
                }
            }
        }
        for key in &["id", "collection_id"] {
            if let Some(id) = created.get(key).and_then(|v| v.as_str()) {
                let s = id.trim();
                if !s.is_empty() {
                    return s.to_string();
                }
            }
        }
    }
    String::new()
}

fn extract_document_id(res: &serde_json::Value) -> String {
    if let Some(data) = res.get("data") {
        for key in &["id", "file_id", "document_id"] {
            if let Some(id) = data.get(key).and_then(|v| v.as_str()) {
                let s = id.trim();
                if !s.is_empty() {
                    return s.to_string();
                }
            }
        }
    }
    for key in &["id", "file_id", "document_id"] {
        if let Some(id) = res.get(key).and_then(|v| v.as_str()) {
            let s = id.trim();
            if !s.is_empty() {
                return s.to_string();
            }
        }
    }
    String::new()
}

fn matches_simple_glob(name: &str, pattern: &str) -> bool {
    if let Some(ext) = pattern.strip_prefix("*.") {
        name.ends_with(&format!(".{ext}"))
    } else {
        name == pattern
    }
}

fn print_help() {
    println!("xAI Collections Knowledge Base");
    println!();
    println!("Usage: xint collections <subcommand> [options]");
    println!();
    println!("Subcommands:");
    println!("  list                                List all collections");
    println!("  create --name <n> [--description <d>]  Create a collection");
    println!("  ensure --name <n>                   Create collection if it doesn't exist");
    println!("  add-document --collection-id <id> --document-id <id>");
    println!("  upload --path <file> [--purpose <p>] Upload a file");
    println!("  search --query <q> [--collection-ids <id1,id2>]");
    println!("  sync-dir --collection-name <n> --dir <path> [--glob *.md] [--limit 50]");
    println!();
    println!("Env vars:");
    println!("  XAI_API_KEY              File upload + document search");
    println!("  XAI_MANAGEMENT_API_KEY   Collections management");
}
